src/Map.ts

```diff
@@ -0,0 +1,261 @@
+import { hasClass } from './decorators';
+import {is} from './object';
+import global from './global';
+
+module Native {
+	interface NativeMap<K, V> {
+		size: number;
+		clear(): void;
+		delete(key: K): boolean;
+		entries(): any;
+		forEach(callback: (key: K, value: V, instance: Map<K, V>) => any, context?: {}): void;
+		get(key: K): V;
+		has(key: K): boolean;
+		keys(): K[];
+		set(key: K, value: V): NativeMap<K, V>;
+		values(): V[];
+	}
+
+	// Since ES6 classes aren't inheritable, create a thin wrapper around
+	// the native WeakMap with an inheritable TS class
+	export class Map<K, V> {
+		private _map: NativeMap<K, V>;
+
+		constructor(iterable?: any) {
```

Comment:
I think this should be able to use `array.ArrayLike` (from `src/array.ts`) instead of `any`.